diff -Nur metacity-2.16.3/src/place.c metacity-2.16.3.new/src/place.c
--- metacity-2.16.3/src/place.c	2006-11-08 04:05:32.000000000 +0100
+++ metacity-2.16.3.new/src/place.c	2006-11-08 04:01:22.000000000 +0100
@@ -858,6 +858,8 @@
   x = xi->rect.x;
   y = xi->rect.y;
 
+  n_xineramas = 1;
+
   if (find_first_fit (window, fgeom, windows,
                       xineramas_list, n_xineramas,
                       x, y, &x, &y))
