diff -Nuard metacity-2.6.3.orig/src/main.c metacity-2.6.3/src/main.c
--- metacity-2.6.3.orig/src/main.c	2003-09-29 18:55:26.000000000 +0200
+++ metacity-2.6.3/src/main.c	2003-10-30 02:33:27.000000000 +0100
@@ -425,13 +425,27 @@
    * don't want users to lose their WM :-/
    */
   if (!meta_ui_have_a_theme ())
-    meta_ui_set_current_theme ("Simple", FALSE);
-  
-  if (!meta_ui_have_a_theme ())
-    meta_ui_set_current_theme ("Atlanta", FALSE);
+  {
+    gchar  *dir_entry = NULL;
+    GError *err = NULL;
+    GDir   *themes_dir = NULL;
+    
+    if (!(themes_dir = g_dir_open (METACITY_DATADIR"/themes",0,&err)))
+      {
+        meta_fatal (_("Failed to scan themes directory : %s\n"), err->message);
+        g_error_free (err);
+      } 
+    else 
+      {
+        while (((dir_entry = g_dir_read_name (themes_dir)) != NULL) && 
+	       (!meta_ui_have_a_theme ()))
+	  {
+            meta_ui_set_current_theme (dir_entry, FALSE);
+          }
 
-  if (!meta_ui_have_a_theme ())
-    meta_ui_set_current_theme ("Crux", FALSE);
+       g_dir_close (themes_dir);
+      }
+   }
 
   if (!meta_ui_have_a_theme ())
     meta_fatal (_("Could not find a theme! Be sure %s exists and contains the usual themes."),
