diff -urN aa/metacity-2.7.0/configure.in metacity-2.7.0/configure.in
--- aa/metacity-2.7.0/configure.in	2004-02-15 23:32:23.000000000 +0100
+++ metacity-2.7.0/configure.in	2004-02-16 01:05:02.074172808 +0100
@@ -134,7 +134,7 @@
 ## try definining HAVE_BACKTRACE
 AC_CHECK_HEADERS(execinfo.h, [AC_CHECK_FUNCS(backtrace)])
 
-ALL_LINGUAS="am ar az be bg bn ca cs cy da de el en_GB es fa fi fr ga gl he hi hr hu id is it ja ko lt lv mk ml mn ms nl nn no pl pt pt_BR ro ru sl sk sq sr sr@Latn sv ta tr uk vi wa zh_CN zh_TW"
+ALL_LINGUAS="am ar az be bg bn ca cs cy da de el en_GB es fa fi fr ga gl he hi hr hu id is it ja ko lt lv mk ml mn ms nb nl nn pl pt pt_BR ro ru sl sk sq sr sr@Latn sv ta tr uk vi wa zh_CN zh_TW"
 AM_GLIB_GNU_GETTEXT
 
 ## here we get the flags we'll actually use
