diff -Nuard metacity-2.6.3.orig/src/main.c metacity-2.6.3/src/main.c
--- metacity-2.6.3.orig/src/main.c	2003-09-29 18:55:26.000000000 +0200
+++ metacity-2.6.3/src/main.c	2003-10-30 00:47:30.000000000 +0100
@@ -426,7 +426,7 @@
    */
   if (!meta_ui_have_a_theme ())
     meta_ui_set_current_theme ("Simple", FALSE);
-  
+
   if (!meta_ui_have_a_theme ())
     meta_ui_set_current_theme ("Atlanta", FALSE);
 
@@ -434,6 +434,24 @@
     meta_ui_set_current_theme ("Crux", FALSE);
 
   if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("AgingGorilla", FALSE);
+
+  if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("Bright", FALSE);
+
+  if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("Esco", FALSE);
+
+  if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("Metabox", FALSE);
+
+  if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("Bluecurve", FALSE);
+
+  if (!meta_ui_have_a_theme ())
+    meta_ui_set_current_theme ("Industrial", FALSE);
+
+  if (!meta_ui_have_a_theme ())
     meta_fatal (_("Could not find a theme! Be sure %s exists and contains the usual themes."),
                 METACITY_DATADIR"/themes");
   
