diff -Nur metacity-2.4.1.orig/src/Makefile.am metacity-2.4.1/src/Makefile.am
--- metacity-2.4.1.orig/src/Makefile.am	Tue Aug 20 22:18:18 2002
+++ metacity-2.4.1/src/Makefile.am	Wed Sep 18 18:37:21 2002
@@ -113,9 +113,6 @@
 
 @INTLTOOL_SCHEMAS_RULE@
 
-install-data-local:
-	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(srcdir)/$(schema_DATA)
-
 IMAGES=default_icon.png stock_maximize.png stock_minimize.png stock_delete.png
 VARIABLES=default_icon_data $(srcdir)/default_icon.png \
           stock_maximize_data $(srcdir)/stock_maximize.png \
