--- metacity-2.3.233/src/Makefile.am~	Fri May  3 18:44:39 2002
+++ metacity-2.3.233/src/Makefile.am	Fri May  3 18:44:39 2002
@@ -106,7 +106,7 @@
 VARIABLES=default_icon_data $(srcdir)/default_icon.png
 
 BUILT_SOURCES = inlinepixbufs.h
-CLEANFILES += inlinepixbufs.h
+CLEANFILES = inlinepixbufs.h
 
 inlinepixbufs.h: $(IMAGES)
 	   $(GDK_PIXBUF_CSOURCE) --raw --build-list $(VARIABLES) >$(srcdir)/inlinepixbufs.h
