--- metacity-2.4.5/src/Makefile.am.wiget	Tue Nov 26 19:26:51 2002
+++ metacity-2.4.5/src/Makefile.am	Tue Nov 26 19:33:02 2002
@@ -31,8 +31,6 @@
 	frame.h					\
 	frames.c				\
 	frames.h				\
-	gradient.c				\
-	gradient.h				\
 	group.c					\
 	group.h					\
 	iconcache.c				\
@@ -61,16 +59,10 @@
 	stack.h					\
 	tabpopup.c				\
 	tabpopup.h				\
-	theme.c					\
-	theme.h					\
-	theme-parser.c				\
-	theme-parser.h				\
 	themewidget.c				\
 	themewidget.h				\
 	ui.c					\
 	ui.h					\
-	util.c					\
-	util.h					\
 	window.c				\
 	window.h				\
 	window-props.c				\
@@ -114,16 +106,16 @@
 libexec_PROGRAMS=metacity-dialog
 
 EFENCE=
-metacity_LDADD= $(EFENCE) @METACITY_LIBS@ $(EFENCE)
+metacity_LDADD= $(EFENCE) @METACITY_LIBS@ $(EFENCE) libmetacity-private.la
 metacity_theme_viewer_LDADD= @METACITY_LIBS@ libmetacity-private.la
 metacity_dialog_LDADD=@METACITY_LIBS@
 
-testgradient_SOURCES=gradient.h gradient.c testgradient.c
+testgradient_SOURCES=testgradient.c
 testasyncgetprop_SOURCES=async-getprop.h async-getprop.c testasyncgetprop.c
 
 noinst_PROGRAMS=testgradient testasyncgetprop
 
-testgradient_LDADD= @METACITY_LIBS@
+testgradient_LDADD= @METACITY_LIBS@ libmetacity-private.la
 testasyncgetprop_LDADD= @METACITY_LIBS@
 
 desktopfilesdir=$(datadir)/gnome/wm-properties
