diff -urN metacity-2.4.8.orig/src/Makefile.am metacity-2.4.8/src/Makefile.am
--- metacity-2.4.8.orig/src/Makefile.am	Mon Dec  9 21:58:58 2002
+++ metacity-2.4.8/src/Makefile.am	Mon Dec  9 22:03:04 2002
@@ -31,8 +31,6 @@
 	frame.h					\
 	frames.c				\
 	frames.h				\
-	gradient.c				\
-	gradient.h				\
 	group.c					\
 	group.h					\
 	group-private.h				\
@@ -64,16 +62,10 @@
 	stack.h					\
 	tabpopup.c				\
 	tabpopup.h				\
-	theme.c					\
-	theme.h					\
-	theme-parser.c				\
-	theme-parser.h				\
 	themewidget.c				\
 	themewidget.h				\
 	ui.c					\
 	ui.h					\
-	util.c					\
-	util.h					\
 	window.c				\
 	window.h				\
 	window-props.c				\
@@ -117,17 +109,17 @@
 libexec_PROGRAMS=metacity-dialog
 
 EFENCE=
-metacity_LDADD=@METACITY_LIBS@ $(EFENCE)
+metacity_LDADD=@METACITY_LIBS@ $(EFENCE) libmetacity-private.la
 metacity_theme_viewer_LDADD= @METACITY_LIBS@ libmetacity-private.la
 metacity_dialog_LDADD=@METACITY_LIBS@
 
-testgradient_SOURCES=gradient.h gradient.c testgradient.c
+testgradient_SOURCES=testgradient.c
 testasyncgetprop_SOURCES=async-getprop.h async-getprop.c testasyncgetprop.c
 
 noinst_PROGRAMS=testgradient testasyncgetprop
 
-testgradient_LDADD= @METACITY_LIBS@
-testasyncgetprop_LDADD= @METACITY_LIBS@
+testgradient_LDADD= @METACITY_LIBS@ libmetacity-private.la
+testasyncgetprop_LDADD= @METACITY_LIBS@ 
 
 desktopfilesdir=$(datadir)/gnome/wm-properties
 desktopfiles_in_files=metacity.desktop.in
